### ISSUE TITLE
Slow training for tf.keras.layers.Embedding with the new optimizers

### ISSUE THREAD

**Issue by vysarge** [keras-team/tf-keras#45]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04.6 LTS
- TensorFlow installed from (source or binary): binary, presumably
- TensorFlow version (use command below): v2.12.0-rc1-12-g0db597d0d75
- Python version: 3.10.12 (main, Jun  7 2023, 12:45:35) [GCC 9.4.0]
- Bazel version (if compiling from source): n/a
- GPU model and memory: NVIDIA Tesla T4 16GB
- Exact command to reproduce: [Code in colab here.](https://colab.research.google.com/drive/1qx5JNlgjIpcdshdAQCpRT3Smxq_etfWm?usp=sharing)

**Describe the problem**.

I'm seeing a significant performance regression when testing training with large embedding tables (tf.keras.layers.Embedding) with some of the new optimizers compared to the legacy optimizers.  Times are logged from the last cell of the colab notebook; for larger tables this can be a severalfold slowdown.  In this tiny test I see the issue with SGD, Adadelta, Adagrad, Adamax, and Ftrl.  Relative to legacy, the new Nadam optimizer performs similarly and the new Adam and RMSProp optimizers speed up.

As a side-note, I'm not certain if this should be filed as a separate issue (and perhaps to XLA directly) but I am also seeing slowdowns when passing `jit_compile=True` to `model.compile` with some of the new optimizers.  In the colab notebook, altering one line:
`model.compile(optimizer=o, loss='binary_crossentropy')`
to
`model.compile(optimizer=o, loss='binary_crossentropy', jit_compile=True)`
in the final cell displays a further slowdown for SGD, Adagrad, and FTRL optimizers, for example.

**Describe the current behavior**.

Rough performance figures summarized from this test (all ms/epoch for the same type of random data):
* SGD: 97.2 legacy vs 605.1 new
* Adadelta: 69.6 legacy vs 842.1 new
* Adagrad: 66.9 legacy vs 88.2 new
* Adamax: 72.7 legacy vs 898.6 new
* FTRL: 71.5 legacy vs 983.2 new

**Standalone code to reproduce the issue**.

[Code in colab here.](https://colab.research.google.com/drive/1qx5JNlgjIpcdshdAQCpRT3Smxq_etfWm?usp=sharing)

**Comment by tilakrayal** [comment]
I tried to execute the mentioned code multiple iterations with the latest keras version and observed that training time was gradually decreased. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/9f134f6ec6b1e2cc5a49ad49ed622388/untitled2402.ipynb).
